packages/playground/data-liberation/src/entity-readers/WP_WXR_Sorted_Entity_Reader.php

```diff
@@ -0,0 +1,669 @@
+		$max_index_length = 191;
+
+		/**
+		 * This is a table used to map the IDs of the imported entities. It is
+		 * used to map all the IDs of the entities.
+		 *
+		 * @param int $id The ID of the entity.
+		 * @param int $session_id The current session ID.
+		 * @param int $entity_type The type of the entity, comment, etc.
+		 * @param string $entity_id The ID of the entity before the import.
+		 * @param string $mapped_id The mapped ID of the entity after import.
+		 * @param string $parent_id The parent ID of the entity.
+		 * @param string $cursor_id The cursor ID of the entity.
+		 * @param int $sort_order The sort order of the entity.
+		 */
+		$sql = $wpdb->prepare(
+			'CREATE TABLE IF NOT EXISTS %i (
+				id bigint(20) unsigned NOT NULL AUTO_INCREMENT,
+				session_id bigint(20) unsigned,
+				entity_type tinyint(1) NOT NULL,
+				entity_id text NOT NULL,
+				mapped_id text DEFAULT NULL,
+				parent_id text DEFAULT NULL,
+				cursor_id text DEFAULT NULL,
+				sort_order int DEFAULT 1,
+				PRIMARY KEY  (id),
+				KEY session_id (session_id),
+				KEY entity_id (entity_id(%d)),
+				KEY parent_id (parent_id(%d)),
+				KEY cursor_id (cursor_id(%d))
+			) ' . $wpdb->get_charset_collate(),
+			self::get_table_name(),
+			$max_index_length,
+			$max_index_length,
+			$max_index_length
+		);
+
+		require_once ABSPATH . 'wp-admin/includes/upgrade.php';
+		// dbDelta is a function that creates the table.
+		dbDelta( $sql );
+	}
+
+	/**
+	 * Run by register_deactivation_hook or similar. It drops the table and
+	 * deletes the option.
+	 */
+	public static function delete_db() {
+		global $wpdb;
+
+		// Drop the table.
+		$wpdb->query(
+			$wpdb->prepare( 'DROP TABLE IF EXISTS %i', self::get_table_name() )
+		);
+	}
+
+	/**
+	 * Set the emit cursor flag. If true, the reader will emit the cursor ID
+	 * for each entity.
+	 *
+	 * @param bool $emit_cursor The emit cursor flag.
+	 */
+	public function set_emit_cursor( $emit_cursor ) {
+		$this->emit_cursor = $emit_cursor;
+	}
+
+	/**
+	 * Reset the class.
+	 */
+	public function reset() {
+		$this->current_session = null;
+	}
+
+	/**
+	 * Delete all rows for a given session ID.
+	 *
+	 * @param int $session_id The session ID to delete rows for.
+	 * @return int|false The number of rows deleted, or false on error.
+	 */
+	public function delete_session( $session_id = null ) {
+		global $wpdb;
+
+		return $wpdb->delete(
+			self::get_table_name(),
+			array( 'session_id' => $session_id ?? $this->current_session ),
+			array( '%d' )
+		);
+	}
+
+	/**
+	 * Add the next entity to the sorting table.
+	 *
+	 * @param string $entity_type The type of the entity.
+	 * @param array  $data The data to map.
+	 * @param mixed  $cursor_id The stream cursor ID.
+	 */
+	public function add_next_entity( $entity = null ) {
+		global $wpdb;
+
+		// We're done if all the entities are processed
+		if ( ! $this->valid() ) {
+			return false;
+		}
+
+		$entity      = $entity ?? $this->current();
```

Comment:
I'm having second thoughts about the "every entity reader is an iterator" idea.

There's `get_entity()` and `current()`. The latter has a name that suggests it's the same as the former, but it doesn't actually do the same thing. The PHP iterator design assumes a newly created iterator is already initialized and `current()` comes before `next()`, which is in direct opposition to the Reader that expects `next_entity()` to be called before `get_entity()`.

I'm thinking we could either remove the iterator-ness completely, or create a single `new Entity_Reader_Iterator( $reader )` class to have a clearly separated single point of entry.